### DISCUSSION TITLE
First Time contribution

### DISCUSSION THREAD

**Discussion by Ahmed-elshorbagy** [invertase/flutterfire_desktop#66] [Q&A] [answered]
I can help to speed up the process .
I can start with cloud firstore or whatever the next step in the roadmap
this is my first time in open source projects
can anyone guide me 
thanks in advance

**Comment by peterBrxwn** [comment]
Same here. Me and my team (3 other developers ~2 yrs experience with flutter, dart and javascript) would love to help. But we have no idea where to start.

**Reply by TimWhiting** [reply]
Glad to see interest in contributing! Happy to help answer questions.

**Comment by TimWhiting** [comment, marked as answer]
Here is the Roadmap.
https://github.com/invertase/flutterfire_desktop/issues/52

I contributed the cloud_functions api a while back and I can share what my opinion is and what I learned, but I'll defer to @pr-Mais who seems to be leading this effort.

@pr-Mais is still working on finishing auth. But the main pieces should be available in core / auth to start working on the next steps.

I would pick one of the simple next step packages listed in the roadmap (copied below) and let @pr-Mais know which one you are working on so that work is not duplicated:
- Installations
- Remote Config
- Cloud Storage
- Analytics

I encourage working on one of the simpler packages that is next in the Roadmap and not tackling something huge like database or firestore for a few reasons:
- Learning how to contribute to this project
- Getting feedback on how you are structuring the code via PRs
- It would be easy to get discouraged trying to work on a massive thing like database or firestore by yourself
- Database and firestore are not the priority in the roadmap and are thus less likely to receive prompt reviews
- The sooner we get these packages done, the sooner more effort and priority can be placed on database and firestore

Next I would look at how the existing projects are structured
- core is simple and doesn't involve external APIs so it wouldn't be helpful
- functions is rather simple as well, and so it is structured a little more simply
- auth is very complex, but as such it has a more structured approach

Functions was simple enough that I got it done in just a couple of PRs

- Try creating one PR for just the basic user facing interfaces and class definitions (leave all the functions throwing Unimplemented errors). Use the fluttefire plugin platform interface package and main plugin package to understand how to structure the user facing API
- Next create a PR with an API class wrapping the internal googleapi calls. (Depending on whether the package has a corresponding api in https://pub.dev/packages/googleapis this may be more or less needed), functions delegates calls to user defined endpoints so I did not have a separate API class when I implemented that. 
- Next start implementing the functions a couple at a time, with corresponding tests. Look at existing tests for inspiration. Create one PR for a couple of functions to make it easier to review the PR. You can use the [javascript sdk](https://github.com/firebase/firebase-js-sdk/) as a reference for implementing the logic.

More detailed steps with links to auth examples:
- Create a dart only package
- Create a flutter desktop project (not a plugin, just a package) that depends on the dart only package & the platform interface package from the flutterfire sdk and mark it as implementing that platform interface in the pubspec.yaml for the desktop platforms [auth example](https://github.com/invertase/flutterfire_desktop/blob/65fad4d7d2c13544709246acb1d833fe916db9b0/packages/firebase_auth/firebase_auth_desktop/pubspec.yaml#L30)
- Create an API that is a light wrapper around the googleapi endpoints [auth example](https://github.com/invertase/flutterfire_desktop/blob/65fad4d7d2c13544709246acb1d833fe916db9b0/packages/firebase_auth/firebase_auth_dart/lib/src/api/api.dart#L164)
- Create a class that mimics the api of the fluttefire version of the plugins [auth example](https://github.com/invertase/flutterfire_desktop/blob/65fad4d7d2c13544709246acb1d833fe916db9b0/packages/firebase_auth/firebase_auth_dart/lib/src/firebase_auth.dart)
- Create a class in the desktop package that implements the flutterfire platform interface of the package and delegates to the dart package for the implementation [auth example](https://github.com/invertase/flutterfire_desktop/blob/65fad4d7d2c13544709246acb1d833fe916db9b0/packages/firebase_auth/firebase_auth_desktop/lib/firebase_auth_desktop.dart)
- You might have to implement other platform interfaces as well depending on the package which should delegate to the dart package classes. [auth example](https://github.com/invertase/flutterfire_desktop/blob/main/packages/firebase_auth/firebase_auth_desktop/lib/src/firebase_auth_user.dart)
- Additionally you'll probably have to map the classes that do not have platform interfaces that you need to implement but are duplicated across the pure-dart and the platform interface.  [auth example](https://github.com/invertase/flutterfire_desktop/blob/main/packages/firebase_auth/firebase_auth_desktop/lib/src/utils/desktop_utils.dart)

**Reply by TimWhiting** [reply]
Additionally, feel free to continue to discuss here any questions you have about contributing in general or open an issue about specific packages when you start working on them, and we can answer specific questions in that issue.

**Reply by Ahmed-elshorbagy** [reply]
thanks for reply 
can you distribute the workload so that our works doesn't get overlapped by each other and avoid time wasting 
thanks in advance

**Reply by TimWhiting** [reply]
@Ahmed-elshorbagy Which of the packages on the Roadmap are you most familiar with from a user's perspective? Pick that one, and comment here as well as open an issue saying that you are planning on contributing that package. If it is complex and a lot of work we can split the work, otherwise I'll have @peterbrown0 and his team work on different plugins. I'm not as familiar with these plugins since I have used mainly database & functions. If you need more help to get started just say so on the issue you open and we can go from there.

**Reply by Ahmed-elshorbagy** [reply]
i am familiar with cloud firestore but since it isn't in the road map i can start with cloud storage

**Reply by TimWhiting** [reply]
You should be able to copy a lot of the platform interface to give you the basic data classes and models that you'll need for the dart package: https://github.com/firebase/flutterfire/tree/master/packages/firebase_storage/firebase_storage_platform_interface, additionally you can take the platform interface folder and copy that into the desktop package and extend those classes. In other words, a lot of the boilerplate code you should be able to reuse, including the implementation of nested path references (in the internal folder of that package I linked).

**Reply by Ahmed-elshorbagy** [reply]
@TimWhiting  i have done some of the implementations but i stuck at some parts can you give me some guidance 
1) don't know if the code files structures are good 
2) don't know if i should import firebase storage platform interface or reimplement it 
this is link to [repository](https://github.com/Ahmed-elshorbagy/flutterfire_desktop/tree/main/packages/firebase_storage)
thanks in advance

**Reply by TimWhiting** [reply]
It's looking good. If you open a draft pull request to indicate that you are working on it and make it easier to provide feedback, that would be nice.

A few initial items of feedback.
- Include the copyright at the top of each of the dart files:
```dart
// Copyright 2022 Invertase Limited. All rights reserved.
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE file.
```
- Copy the LICENSE from one of the other packages.
- The first versions of these packages have typically been 0.1.0-dev.0 @pr-Mais is that still what you want to be doing?
- Look into functions or auth packages to get a basic description template for putting in the pubspec.yaml meta /description information.
- The pure dart package files should not have a `Platform` / `_delegate` (because the dart code should work on all platforms)
For example
```dart
class Reference {
  Reference._(this.storage, this._delegate) {
    ReferencePlatform.verifyExtends(_delegate);
  }

  ReferencePlatform _delegate;
...
```
should be:
```dart
class Reference {
  Reference._(this.storage);
...
```
There is a couple of places this needs to change.

In the desktop package you will need to extend the platform interfaces though: (Yes, import the platform interface package in the desktop package)
```dart
class FirebaseDesktopReference extends ReferencePlatform {
  /// Creates a [ReferencePlatform] that is implemented using pure dart.
  FirebaseDesktopReference(FirebaseStoragePlatform storage, String path)
      : super(storage, path);
...
```
See the implementation of the platform interfaces using a [method channel](https://github.com/firebase/flutterfire/tree/master/packages/firebase_storage/firebase_storage_platform_interface/lib/src/method_channel), you will be essentially doing this, but connecting it to the pure dart package.

- You can copy the example app from the flutterfire repository to test locally the desktop implementation. Just add a path dependency on the desktop package. I believe flutter will automatically register the desktop implementation of plugin for you when you build for desktop.

I hope that answers your questions

**Reply by pr-Mais** [reply]
Hello, thanks @TimWhiting on the amazing instructions, everything mentioned is correct.
Regarding the version, once the work is in a good shape for a preview release, the first release will have the version number `0.1.0-dev.0`.

Further feedback and discussion will be held on the PR, thanks @Ahmed-elshorbagy for initiating it 😄

**Reply by TimWhiting** [reply]
Thanks for the motivation to get involved in this project again. It is good to see other open source contributors actively involved! I've started firebase_remote_config, and will developing that PR as I find time for it.

**Reply by Ahmed-elshorbagy** [reply]
thansks @pr-Mais @TimWhiting  i will update my pr with new files today i hope you can give feedback i intend to finish the whole project by the end of next week

**Reply by TimWhiting** [reply]
By the way there is also https://github.com/invertase/dart_firebase_apis that @Salakar has generated for the firebase apis based on the google discovery documents for those APIs. They do not cover the whole API surface that the javascript sdk covers from what I saw when I glanced at it, but they cover a decent bit depending on the firebase package, and could be a good place to start when working on the API interface portion.
